File: backend/src/settings/lrembecki.core.settings/Companies/CompanyVM.cs
using lrembecki.core.settings.Address;

namespace lrembecki.core.settings.Companies;

public record CompanyVM(
    Guid Id,
    string Name,
    Guid AddressId,
    AddressVM Address,
    List<CompanyParameterVM> Parameters,
    List<Guid> PhoneContact,
    List<Guid> EmailContact)
{
    internal static CompanyVM Map(CompanyEntity company)
    {
        if (company == null) return null!;
        return new(
            Id: company.Id,
            Name: company.Name,
            AddressId: company.AddressId,
            Address: AddressVM.Map(company.Address),
            Parameters: [.. company.Parameters.Select(CompanyParameterVM.Map)],
            PhoneContact: [.. company.PhoneContact.Select(e => e.Id)],
            EmailContact: [.. company.EmailContact.Select(e => e.Id)]
        );
    }
}

File: frontend/src/app/modules/settings/companies/company.vm.ts
export class CompanyParameterVM {
  order: number = null!;
  name: string = null!;
  value: string = null!;

  constructor(init?: Partial<CompanyParameterVM>) {
    Object.assign(this, init);
  }
}

export class AddressVM {
  id: string = null!;
  address1: string = null!;
  address2: string = null!;
  city: string = null!;
  postalCode: string = null!;
  country: string = null!;

  constructor(init?: Partial<AddressVM>) {
    Object.assign(this, init);
  }
}

export class CompanyVM {
  id: string = null!;
  name: string = null!;
  addressId: string = null!;
  address: AddressVM = new AddressVM();
  parameters: CompanyParameterVM[] = [];
  emailContact: string[] = [];
  phoneContact: string[] = [];

  constructor(init?: Partial<CompanyVM>) {
    Object.assign(this, init);
    this.address = new AddressVM(init?.address);
    this.parameters =
      init?.parameters?.map((param) => new CompanyParameterVM(param)) ?? [];
  }
}
